infra/pubsub/rabbitmq/broker.go

```diff
@@ -0,0 +1,248 @@
+package rabbitmq
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/rabbitmq/amqp091-go"
+)
+
+var (
+	ErrConnectionFailed       = errors.New("rabbitmq connection failed")
+	ErrConnectionNotAvailable = errors.New("rabbitmq connection not available")
+	ErrDeclarationFailed      = errors.New("rabbitmq AMQP entity declaration failed")
+)
+
+type Broker interface {
+	Channel(ctx context.Context) (*amqp091.Channel, error)
+	DeclareExchange(ctx context.Context, cfg *ExchangeConfig) error
+	DeclareQueue(ctx context.Context, cfg *QueueConfig, exchange *ExchangeConfig, routingKey string) error
+	Close() error
+}
+
+var _ Broker = (*Connection)(nil)
+
+type Connection struct {
+	cfg         *Config
+	conn        *amqp091.Connection
+	ch          *amqp091.Channel
+	mu          sync.RWMutex
+	done        chan struct{}
+	reconnectCh chan struct{}
+	logger      Logger
+}
+
+func NewConnectionBroker(cfg *Config, logger Logger) (*Connection, error) {
+	if logger == nil {
+		// by default no-operation logger
+		logger = &NoopLogger{}
+	}
+
+	b := &Connection{
+		cfg:         cfg,
+		done:        make(chan struct{}),
+		reconnectCh: make(chan struct{}, 1),
+		logger:      logger,
+	}
+
+	if err := b.connect(); err != nil {
+		return nil, fmt.Errorf("broker creation: %w", err)
+	}
+
+	go b.connectionWatcher()
+	return b, nil
+}
+
+func (b *Connection) connect() error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	conn, err := amqp091.DialConfig(b.cfg.URL, amqp091.Config{
+		Dial: amqp091.DefaultDial(b.cfg.ConnectTimeout),
+	})
+	if err != nil {
+		return fmt.Errorf("%w: %v", ErrConnectionFailed, err)
+	}
+
+	channel, err := conn.Channel()
+	if err != nil {
+		_ = conn.Close()
+		return fmt.Errorf("failed to open channel after connection: %w", err)
```

Comment:
варто переглянути всі помилки, які повертаєш
```suggestion
		return fmt.Errorf("open channel after connection: %w", err)
```

https://github.com/uber-go/guide/blob/master/style.md#errors

When adding context to returned errors, keep the context succinct by avoiding phrases like "failed to", which state the obvious and pile up as the error percolates up through the stack